File: concepts/04_collections_and_iteration/01_iteration_protocol/rust/test_02_iterable_iterator_and_fallbacks.rs
// polyglot-family: collections_and_iteration
// polyglot-concept: iteration_protocol
// polyglot-related: languages/rust/tests/language/05_functions_closures_and_iterators/
// polyglot-related+: test_040_iterator_capabilities_and_early_termination.rs
//
// 共同问题：iterable 与 iterator 是否分离；缺少正式协议时会不会回退到索引；重复遍历是否独立。
// 对照观察：Rust 显式区分 `IntoIterator` 与 `Iterator`，不会把只实现 Index 的类型当 iterable。

use polyglot_rust_course::assert_compile_fails;

#[test]
fn comparison() {
    let range = 0..3;
    assert_eq!(range.clone().collect::<Vec<_>>(), [0, 1, 2]);
    assert_eq!(range.collect::<Vec<_>>(), [0, 1, 2]);
    assert_compile_fails(
        "use std::ops::Index; struct OnlyIndex; impl Index<usize> for OnlyIndex { \
         type Output=i32; fn index(&self, _:usize)->&i32 { &1 } } \
         fn main() { for _ in OnlyIndex {} }",
        &["is not an iterator"],
    );
}

File: concepts/04_collections_and_iteration/01_iteration_protocol/julia/test_02_iterable_iterator_and_fallbacks.jl
# polyglot-family: collections_and_iteration
# polyglot-concept: iteration_protocol
# polyglot-related: languages/julia/language/06_arrays_indexing_iteration_and_broadcast/
# polyglot-related+: test_045_custom_iteration_protocol.jl
#
# 共同问题：iterable 与 iterator 是否必须分离；一次性消费如何显式表达。
# 对照观察：普通 Julia iterable 可重复调用 iterate；Iterators.Stateful 把外部状态包装成可继续消费对象。

using Test

@testset "Stateful 显式保存消费位置" begin
    stateful = Iterators.Stateful(1:3)
    @test popfirst!(stateful) == 1
    @test collect(stateful) == [2, 3]
    @test isempty(stateful)
    @test collect(1:3) == collect(1:3)
end
